Tests/ClientTests/Web Filters/DebouncingResourceDownloaderTests.swift

```diff
@@ -127,125 +139,350 @@ class DebouncingResourceDownloaderTests: XCTestCase {
 
     // Then
-    XCTAssertEqual(returnedURL1, landingURL)
-    XCTAssertNil(returnedURL2)
+    XCTAssertEqual(redirectChain1.last?.url, landingURL)
+    XCTAssertNil(redirectChain2.last)
   }
+  
+  /// Test regex rule where `prepend_scheme` is specified
+  func testRegexRuleWithPrependScheme() {
+    // Given
+    // A constructed url
+    let baseURL = URL(string: "https://brave-scheme.example")!
+    let startURL1 = baseURL.appendingPathComponent("braveattentiontoken.example")
+    let startURL2 = baseURL.appendingPathComponent("braveattentiontoken.example/")
+    let startURL3 = baseURL.appendingPathComponent("://braveattentiontoken.example")
+    let startURL4 = baseURL.appendingPathComponent("https://braveattentiontoken.example")
+    let startURL5 = baseURL.appendingPathComponent("https://braveattentiontoken.example/")
+    let startURL6 = baseURL.appendingPathComponent("https://brave-scheme.example/")
+    let landingURL = URL(string: "https://braveattentiontoken.example")!
+    let landingURL2 = URL(string: "https://braveattentiontoken.example/")!
+
+    // When
+    // A redirect is returned
+    let redirectChain1 = downloader.redirectChain(for: startURL1)
+    let redirectChain2 = downloader.redirectChain(for: startURL2)
+    let redirectChain3 = downloader.redirectChain(for: startURL3)
+    let redirectChain4 = downloader.redirectChain(for: startURL4)
+    let redirectChain5 = downloader.redirectChain(for: startURL5)
+    let redirectChain6 = downloader.redirectChain(for: startURL6)
+
+    // Then
+    XCTAssertEqual(redirectChain1.last?.url, landingURL)
+    XCTAssertEqual(redirectChain2.last?.url, landingURL2)
+    
+    XCTAssertNil(
+      redirectChain3.last?.url,
+      "This should be nil because scheme is not there but we don't want to support ://some-url types of schemeless urls"
+    )
+    XCTAssertNil(
+      redirectChain4.last,
+      "This should be nil because a scheme is already on the raw value and the rule specifies a `prepend_scheme`"
+    )
+    XCTAssertNil(
+      redirectChain5.last,
+      "This should be nil because a scheme is already on the raw value and the rule specifies a `prepend_scheme`"
+    )
+    XCTAssertNil(
+      redirectChain6.last,
+      "This should return nil because the host of the source and destination urls is the same"
+    )
+  }
+  
+  /// Test regex rule where `prepend_scheme` is not specified
+  func testRegexRuleWithNoPrependScheme() {
+    // Given
+    // A constructed url
+    let baseURL = URL(string: "https://brave.com")!
+    let startURL1 = baseURL.appendingPathComponent("https://braveattentiontoken.example")
+    let startURL2 = baseURL.appendingPathComponent("https://braveattentiontoken.example".addingPercentEncoding(withAllowedCharacters: .URLAllowed)!)
+    let startURL3 = baseURL.appendingPathComponent("http://braveattentiontoken.example/")
+    let startURL4 = baseURL.appendingPathComponent("http://192.168.0.1")
```

Comment:
I'd suggest using a non-local IP address just in case we want to exclude these in the future. So maybe `1.1.1.1`.